src/operator/math_functions-inl.h

```diff
@@ -136,6 +118,25 @@ MXNET_BINARY_MATH_FUNC(hypot)
 
 MXNET_BINARY_MATH_FUNC(pow)
 
+template<typename DType> MSHADOW_XINLINE
+float id(DType a) {
```

Comment:
float return for any type? what about half_t and quantization?

Reply:
OK, this looks a bit weird, but the reason is this: We compute math functions either in float or double, because we only have code for these. But some of their gradients are just arithmetic expressions, like x*x + 1. In order to get things correct, I also compute these expressions in float or double. This is why I need id and sqr.

But again, this is cast back to DType in mshadow_op.h.

One thing: If for a special type, you cannot cast back/to float, then none of these math functions will be supported. This is the case for the half2_t type. This is really bad practice, anything nonlinear will fail for this one. The cast should really be implemented.

Reply:
What is the use-case that currently breaks? quantization? half_t? Is it just by inspection, or it is being witnessed?

Reply:
Check div_grad and div_rgrad in mshadow_op.h, for the type mshadow::half::half2_t.
Since cast from that to float does not work, I need the specialization there. If you remove this specialization, it does not build.

It also means that none of the math functions will work for that half2_t type, which seems in any case not completely implemented.